File: asherah-dotnet-shared/src/AppEncryption/Persistence/DynamoDbMetastoreImpl.cs
using System;
using Amazon.DynamoDBv2;
using Amazon.Runtime;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json.Linq;
using LanguageExt;

namespace GoDaddy.Asherah.AppEncryption.Persistence;

public class DynamoDbMetastoreImpl : IMetastore<JObject>
{
    internal DynamoDbMetastoreImpl(Builder builder)
    {
        PreferredRegion = builder.PreferredRegion;
        HasKeySuffix = builder.HasKeySuffix;
        TableName = builder.TableName;
        Endpoint = builder.Endpoint;
        Region = builder.Region;
        Credentials = builder.Credentials;
        DbClient = builder.DbClient;
        Logger = builder.Logger;
    }

    internal string PreferredRegion { get; }
    internal bool HasKeySuffix { get; }
    internal string TableName { get; }
    internal string? Endpoint { get; }
    internal string? Region { get; }
    internal AWSCredentials? Credentials { get; }
    internal IAmazonDynamoDB? DbClient { get; }
    internal ILogger? Logger { get; }

    public static Builder NewBuilder(string region) => new(region);

    public Option<JObject> Load(string keyId, DateTimeOffset created) =>
        throw new NotSupportedException("DynamoDbMetastoreImpl is configuration-only when using native core");

    public Option<JObject> LoadLatest(string keyId) =>
        throw new NotSupportedException("DynamoDbMetastoreImpl is configuration-only when using native core");

    public bool Store(string keyId, DateTimeOffset created, JObject value) =>
        throw new NotSupportedException("DynamoDbMetastoreImpl is configuration-only when using native core");

    public string GetKeySuffix() => HasKeySuffix ? PreferredRegion : string.Empty;

    public interface IBuildStep
    {
        IBuildStep WithKeySuffix();
        IBuildStep WithTableName(string tableName);
        IBuildStep WithCredentials(AWSCredentials credentials);
        IBuildStep WithEndPointConfiguration(string endPoint, string signingRegion);
        IBuildStep WithRegion(string region);
        IBuildStep WithLogger(ILogger logger);
        IBuildStep WithDynamoDbClient(IAmazonDynamoDB client);
        DynamoDbMetastoreImpl Build();
    }

    public class Builder : IBuildStep
    {
        private const string DefaultTableName = "EncryptionKey";

        internal string PreferredRegion { get; }
        internal bool HasKeySuffix { get; private set; }
        internal string TableName { get; private set; } = DefaultTableName;
        internal string? Endpoint { get; private set; }
        internal string? Region { get; private set; }
        internal AWSCredentials? Credentials { get; private set; }
        internal IAmazonDynamoDB? DbClient { get; private set; }
        internal ILogger? Logger { get; private set; }

        public Builder(string region)
        {
            PreferredRegion = region;
        }

        public IBuildStep WithKeySuffix()
        {
            HasKeySuffix = true;
            return this;
        }

        public IBuildStep WithTableName(string tableName)
        {
            TableName = tableName;
            return this;
        }

        public IBuildStep WithCredentials(AWSCredentials credentials)
        {
            Credentials = credentials;
            return this;
        }

        public IBuildStep WithEndPointConfiguration(string endPoint, string signingRegion)
        {
            Endpoint = endPoint;
            Region = signingRegion;
            return this;
        }

        public IBuildStep WithRegion(string region)
        {
            Region = region;
            return this;
        }

        public IBuildStep WithLogger(ILogger logger)
        {
            Logger = logger;
            return this;
        }

        public IBuildStep WithDynamoDbClient(IAmazonDynamoDB client)
        {
            DbClient = client;
            return this;
        }

        public DynamoDbMetastoreImpl Build() => new(this);
    }
}

File: asherah-java/java/src/main/java/com/godaddy/asherah/appencryption/persistence/DynamoDbMetastoreImpl.java
package com.godaddy.asherah.appencryption.persistence;

import com.godaddy.asherah.jni.AsherahConfig;
import org.json.JSONObject;

import java.time.Instant;
import java.util.Optional;

/**
 * DynamoDB metastore adapter. Maps to metastore="dynamodb" in the native config.
 * Compatible with the canonical godaddy/asherah DynamoDbMetastoreImpl.
 */
public class DynamoDbMetastoreImpl implements Metastore<JSONObject> {

    static final String DEFAULT_TABLE_NAME = "EncryptionKey";

    private final String region;
    private final String tableName;
    private final String endPoint;
    private final String signingRegion;
    private final boolean keySuffix;

    private DynamoDbMetastoreImpl(final Builder builder) {
        this.region = builder.region;
        this.tableName = builder.tableName;
        this.endPoint = builder.endPoint;
        this.signingRegion = builder.signingRegion;
        this.keySuffix = builder.keySuffix;
    }

    public void applyConfig(final AsherahConfig.Builder builder) {
        builder.metastore("dynamodb");
        if (region != null) {
            builder.dynamoDbRegion(region);
        }
        if (signingRegion != null) {
            builder.dynamoDbSigningRegion(signingRegion);
        }
        if (tableName != null && !DEFAULT_TABLE_NAME.equals(tableName)) {
            builder.dynamoDbTableName(tableName);
        }
        if (endPoint != null) {
            builder.dynamoDbEndpoint(endPoint);
        }
        if (keySuffix) {
            builder.enableRegionSuffix(true);
        }
    }

    @Override
    public String getKeySuffix() {
        return keySuffix ? "_" + region : "";
    }

    public static Builder newBuilder(final String region) {
        return new Builder(region);
    }

    @Override
    public Optional<JSONObject> load(final String keyId, final Instant created) {
        throw new UnsupportedOperationException("Metastore operations are handled by the native layer");
    }

    @Override
    public Optional<JSONObject> loadLatest(final String keyId) {
        throw new UnsupportedOperationException("Metastore operations are handled by the native layer");
    }

    @Override
    public boolean store(final String keyId, final Instant created, final JSONObject value) {
        throw new UnsupportedOperationException("Metastore operations are handled by the native layer");
    }

    public interface BuildStep {
        BuildStep withTableName(String tableName);
        BuildStep withEndPointConfiguration(String endPoint, String signingRegion);
        BuildStep withKeySuffix();
        DynamoDbMetastoreImpl build();
    }

    public static final class Builder implements BuildStep {
        private final String region;
        private String tableName = DEFAULT_TABLE_NAME;
        private String endPoint;
        private String signingRegion;
        private boolean keySuffix;

        private Builder(final String region) {
            this.region = region;
        }

        @Override
        public BuildStep withTableName(final String tableName) {
            this.tableName = tableName;
            return this;
        }

        @Override
        public BuildStep withEndPointConfiguration(final String endPoint, final String signingRegion) {
            this.endPoint = endPoint;
            this.signingRegion = signingRegion;
            return this;
        }

        @Override
        public BuildStep withKeySuffix() {
            this.keySuffix = true;
            return this;
        }

        @Override
        public DynamoDbMetastoreImpl build() {
            return new DynamoDbMetastoreImpl(this);
        }
    }
}
